File: hub/src/lib/intelligence.js
function asDate(v) {
  const d = v instanceof Date ? v : new Date(v);
  return Number.isNaN(d.getTime()) ? null : d;
}

export function computeThreatLevel(banTimes) {
  const now = Date.now();
  const hour = 3600 * 1000;
  const minute = 60 * 1000;
  const bansHr = banTimes.filter((t) => {
    const d = asDate(t);
    return d && now - d.getTime() < hour;
  }).length;
  const lastMin = banTimes.filter((t) => {
    const d = asDate(t);
    return d && now - d.getTime() < minute;
  }).length;

  if (bansHr === 0) return { level: "LOW", bans_hr: 0, bans_last_min: 0, recommendation: "Trafic normal. Nicio acțiune necesară." };
  if (bansHr < 10) return { level: "LOW", bans_hr: bansHr, bans_last_min: lastMin, recommendation: "Activitate normală de scanare. Monitorizare standard." };
  if (bansHr < 50) return { level: "MEDIUM", bans_hr: bansHr, bans_last_min: lastMin, recommendation: "Activitate suspectă detectată. Verificați jailurile active." };
  if (bansHr < 200) return { level: "HIGH", bans_hr: bansHr, bans_last_min: lastMin, recommendation: "Atac în desfășurare! Considerați blocarea /24 sau CSF." };
  return { level: "CRITICAL", bans_hr: bansHr, bans_last_min: lastMin, recommendation: "Atac DDoS masiv! Activați modul CSFDENY sau contactați datacenter." };
}

export function computeTopAttackers(bans, limit = 10) {
  const m = new Map();
  for (const b of bans) {
    const key = b.ip;
    if (!m.has(key)) m.set(key, { count: 0, jails: new Set(), country: "", country_code: "", isp: "" });
    const e = m.get(key);
    e.count += 1;
    e.jails.add(b.jail);
    if (b.country) {
      e.country = b.country;
      e.country_code = b.country_code || "";
      e.isp = b.isp || "";
    }
  }
  return [...m.entries()].sort((a, b) => b[1].count - a[1].count).slice(0, limit).map(([ip, d]) => ({
    ip,
    count: d.count,
    jails: [...d.jails],
    country: d.country,
    country_code: d.country_code,
    isp: d.isp
  }));
}

export function computeJailStats(bans) {
  const m = new Map();
  for (const b of bans) m.set(b.jail, (m.get(b.jail) || 0) + 1);
  return [...m.entries()].sort((a, b) => b[1] - a[1]).map(([jail, count]) => ({ jail, count }));
}

export function computeCountryStats(bans) {
  const m = new Map();
  for (const b of bans) {
    const code = b.country_code || "XX";
    if (!m.has(code)) m.set(code, { count: 0, country: "", lat: 0, lon: 0, ip: new Map(), jail: new Map(), isp: new Map() });
    const e = m.get(code);
    e.count += 1;
    e.country = b.country || e.country;
    if (b.lat) {
      e.lat = b.lat;
      e.lon = b.lon;
    }
    if (b.ip) e.ip.set(b.ip, (e.ip.get(b.ip) || 0) + 1);
    if (b.jail) e.jail.set(b.jail, (e.jail.get(b.jail) || 0) + 1);
    if (b.isp) e.isp.set(b.isp, (e.isp.get(b.isp) || 0) + 1);
  }
  return [...m.entries()].sort((a, b) => b[1].count - a[1].count).map(([code, d]) => ({
    code,
    country: d.country,
    count: d.count,
    lat: d.lat,
    lon: d.lon,
    unique_ips: d.ip.size,
    ips: [...d.ip.entries()].sort((a, b) => b[1] - a[1]).slice(0, 30).map(([ip, count]) => ({ ip, count })),
    jails: [...d.jail.entries()].sort((a, b) => b[1] - a[1]).slice(0, 8).map(([name, count]) => ({ name, count })),
    isps: [...d.isp.entries()].sort((a, b) => b[1] - a[1]).slice(0, 5).map(([name, count]) => ({ name, count }))
  }));
}

export function computeBanTimeline(bans) {
  const now = new Date();
  const buckets = new Map();
  for (const b of bans) {
    const d = asDate(b.ts);
    if (!d) continue;
    if (now.getTime() - d.getTime() <= 24 * 3600 * 1000) {
      const h = `${String(d.getHours()).padStart(2, "0")}:00`;
      buckets.set(h, (buckets.get(h) || 0) + 1);
    }
  }
  const out = [];
  for (let i = 23; i >= 0; i -= 1) {
    const t = new Date(now.getTime() - i * 3600 * 1000);
    const h = `${String(t.getHours()).padStart(2, "0")}:00`;
    out.push({ hour: h, count: buckets.get(h) || 0 });
  }
  return out;
}

File: backend/intelligence.py
"""Calcule threat intelligence din liste de bannuri."""

from collections import defaultdict
from datetime import datetime, timedelta


def compute_threat_level(ban_times):
    now = datetime.utcnow()
    bans_hr = sum(1 for t in ban_times if (now - t).total_seconds() < 3600)
    last_min = sum(1 for t in ban_times if (now - t).total_seconds() < 60)
    if bans_hr == 0:
        level, rec = "LOW", "Trafic normal. Nicio acțiune necesară."
    elif bans_hr < 10:
        level, rec = "LOW", "Activitate normală de scanare. Monitorizare standard."
    elif bans_hr < 50:
        level, rec = "MEDIUM", "Activitate suspectă detectată. Verificați jailurile active."
    elif bans_hr < 200:
        level, rec = "HIGH", "Atac în desfășurare! Considerați blocarea /24 sau CSF."
    else:
        level, rec = "CRITICAL", "Atac DDoS masiv! Activați modul CSFDENY sau contactați datacenter."
    return {"level": level, "bans_hr": bans_hr, "bans_last_min": last_min, "recommendation": rec}


def compute_top_attackers(bans, limit=10):
    counts = defaultdict(lambda: {"count": 0, "jails": set(), "country": "", "country_code": "", "isp": ""})
    for b in bans:
        ip = b["ip"]
        counts[ip]["count"] += 1
        counts[ip]["jails"].add(b["jail"])
        if b.get("country"):
            counts[ip]["country"] = b["country"]
            counts[ip]["country_code"] = b.get("country_code", "")
            counts[ip]["isp"] = b.get("isp", "")
    return [
        {"ip": ip, "count": d["count"], "jails": list(d["jails"]),
         "country": d["country"], "country_code": d["country_code"], "isp": d["isp"]}
        for ip, d in sorted(counts.items(), key=lambda x: x[1]["count"], reverse=True)[:limit]
    ]


def compute_country_stats(bans):
    counts = defaultdict(lambda: {
        "count": 0, "country": "", "lat": 0, "lon": 0, "code": "",
        "ip_counts": defaultdict(int),
        "jail_counts": defaultdict(int),
        "isp_counts": defaultdict(int),
    })
    for b in bans:
        cc = b.get("country_code") or "XX"
        entry = counts[cc]
        entry["count"] += 1
        if b.get("country"):
            entry["country"] = b["country"]
            entry["code"] = cc
        if b.get("lat"):
            entry["lat"] = b["lat"]
            entry["lon"] = b["lon"]
        ip = b.get("ip")
        if ip:
            entry["ip_counts"][ip] += 1
        jail = b.get("jail")
        if jail:
            entry["jail_counts"][jail] += 1
        isp = b.get("isp")
        if isp:
            entry["isp_counts"][isp] += 1
    return [
        {
            "code": cc,
            "country": d["country"],
            "count": d["count"],
            "lat": d["lat"],
            "lon": d["lon"],
            "unique_ips": len(d["ip_counts"]),
            "ips": [
                {"ip": ip, "count": c}
                for ip, c in sorted(d["ip_counts"].items(), key=lambda x: x[1], reverse=True)[:30]
            ],
            "jails": [
                {"name": j, "count": c}
                for j, c in sorted(d["jail_counts"].items(), key=lambda x: x[1], reverse=True)[:8]
            ],
            "isps": [
                {"name": isp, "count": c}
                for isp, c in sorted(d["isp_counts"].items(), key=lambda x: x[1], reverse=True)[:5]
            ],
        }
        for cc, d in sorted(counts.items(), key=lambda x: x[1]["count"], reverse=True)
    ]


def compute_jail_stats(bans):
    counts = defaultdict(int)
    for b in bans:
        counts[b["jail"]] += 1
    return [{"jail": j, "count": c} for j, c in sorted(counts.items(), key=lambda x: x[1], reverse=True)]


def compute_ban_timeline(bans):
    now = datetime.utcnow()
    buckets = defaultdict(int)
    for b in bans:
        try:
            ts = b["ts"] if isinstance(b["ts"], datetime) else datetime.fromisoformat(str(b["ts"]))
            if (now - ts).total_seconds() <= 86400:
                buckets[ts.strftime("%H:00")] += 1
        except Exception:
            pass
    result = []
    for i in range(23, -1, -1):
        t = now - timedelta(hours=i)
        result.append({"hour": t.strftime("%H:00"), "count": buckets.get(t.strftime("%H:00"), 0)})
    return result
